tensorflow_datasets/core/utils/gpath.py

```diff
@@ -0,0 +1,126 @@
+"""GcsPath wrapper around the gfile API"""
+
+import os
+import pathlib
+import typing
+from typing import Any, AnyStr, Iterator, Optional, Union
+
+import tensorflow as tf
+
+from tensorflow_datasets.core.utils import type_utils
+
+PathArg = Union[str, pathlib.Path, 'GcsPath']
+
+
+class GcsPath(pathlib.PurePosixPath, type_utils.ReadWritePath):
+  """Pathlib like api of tf.io.gfile"""
+  def __new__(cls, *parts: type_utils.PathLike):
+    full_path = '/'.join(os.fspath(p) for p in parts)
+    if full_path.startswith(('gs://')):
+      return super().__new__(cls, full_path.replace('gs://', '/gs/', 1))
+    if full_path.startswith(('/gs/')):
+      return super().__new__(cls, *parts)
+    else:
+      raise ValueError('Invalid path')
+
+
+  def _get_path_str(self) -> str:
+    value = super().__str__()
+    if value.startswith('/gs/'):
+      return value.replace('/gs/', 'gs://', 1)
+    return value  # TODO: Could cache the value
+
+
+  def __fpath__(self) -> str:
+    return self._get_path_str()
+
+
+  def __str__(self) -> str:
+    return self._get_path_str()
+
+
+  def exists(self) -> bool:
+    """Returns True if self exists."""
+    return tf.io.gfile.exists(str(self))
+
+
+  def iterdir(self) -> Iterator['GcsPath']:
+    """Iterates over the directory."""
+    for f in tf.io.gfile.listdir(str(self)):
+      yield GcsPath(f)
+
+
+  def is_dir(self) -> bool:
+    """Returns True if self is a directory."""
+    return tf.io.gfile.isdir(str(self))
+
+
+  def glob(self, pattern: str) -> Iterator['GcsPath']:
+    """Yielding all matching files (of any kind)."""
+    path = os.path.join(str(self), pattern)
+    for f in tf.io.gfile.glob(str(path)):
```

Comment:
Rather than using `str`, it's best to use `_get_path_str()` everywhere, as it makes it easier to inherit the class.

Reply:
Done